src/sinks/aws_s3/config.rs

```diff
@@ -35,29 +35,83 @@ const DEFAULT_KEY_PREFIX: &str = "date=%F/";
 const DEFAULT_FILENAME_TIME_FORMAT: &str = "%s";
 const DEFAULT_FILENAME_APPEND_UUID: bool = true;
 
-#[derive(Deserialize, Serialize, Debug, Clone)]
-#[serde(deny_unknown_fields)]
+/// Configuration for the `aws_s3` sink.
+#[configurable_component(sink)]
+#[derive(Clone, Debug)]
 pub struct S3SinkConfig {
+    /// The S3 bucket name.
+    ///
+    /// This must not include a leading `s3://` or a trailing `/`.
     pub bucket: String,
+
+    /// A prefix to apply to all object keys.
+    ///
+    /// Prefixes are useful for partitioning objects, such as by creating an object key that
+    /// stores objects under a particular "directory". If using a prefix for this purpose, it must end
+    /// in `/` in order to act as a directory path: Vector will **not** add a trailing `/` automatically.
+    #[configurable(metadata(templateable))]
     pub key_prefix: Option<String>,
+
+    /// The timestamp format for the time component of the object key.
+    ///
+    /// By default, object keys are appended with a timestamp that reflects when the objects are
+    /// sent to S3, such that the resulting object key is functionally equivalent to joining the key
+    /// prefix with the formatted timestamp, such as `date=2022-07-18/1658176486`.
+    ///
+    /// This would represent a `key_prefix` set to `date=%F/` and the timestamp of Mon Jul 18 2022
+    /// 20:34:44 GMT+0000, with the `filename_time_format` being set to `%s`, which renders
+    /// timestamps in seconds since the Unix epoch.
+    ///
+    /// Supports the common [`strftime`][chrono_strftime_specifiers] specifiers found in most
+    /// languages.
+    ///
+    /// When set to an empty string, no timestamp will be appended to the key prefix.
+    ///
+    /// [chrono_strftime_specifiers]: https://docs.rs/chrono/latest/chrono/format/strftime/index.html#specifiers
     pub filename_time_format: Option<String>,
+
+    /// Whether or not to append a UUID v4 token to the end of the object key.
+    ///
+    /// The UUID is appended to the timestamp portion of the object key, such that if the object key
+    /// being generated was `date=2022-07-18/1658176486`, setting this field to `true` would result
+    /// in an object key that looked like `date=2022-07-18/1658176486-30f6652c-71da-4f9f-800d-a1189c47c547`.
+    ///
+    /// This ensures there are no name collisions, and can be useful in high-volume workloads where
+    /// object keys must be unique.
     pub filename_append_uuid: Option<bool>,
+
+    /// The filename extension to use in the object key.
     pub filename_extension: Option<String>,
+
     #[serde(flatten)]
     pub options: S3Options,
+
     #[serde(flatten)]
     pub region: RegionOrEndpoint,
+
     #[serde(flatten)]
     pub encoding: EncodingConfigWithFraming,
```

Comment:
I would have expected these to need a `#[configurable(derived)]`?

Reply:
When fields are flattened, we're replacing this one field with N fields in the actual configuration file, so we don't need a title/description for `options` or `S3Options`... so long as each of the fields on `S3Options` have a title/description (or are derived, etc...).

Reply:
AH, doh - yeah that makes sense. Silly in hindsight, thanks for explanation.